pandas/io/parsers.py

```diff
@@ -786,17 +786,25 @@ def read(self, nrows=None):
             return ret
 
         # May alter columns / col_dict
-        index, columns, col_dict = self._create_index(ret)
+        index, columns, mappings, col_dict = self._create_index(ret)
 
-        df = DataFrame(col_dict, columns=columns, index=index)
+        if not mappings:
```

Comment:
I agree, but at this point, there are no obvious simplifications like there were with `tokenizer.c` for example.  And at that point, I prefer the correctness.  Also, it's probably not going to be a **single** loop over the columns.  Secondly, that's not the performance hit I am worried about.  The performance hit will come from when you have to do all of the name changes, and you find yourself having to do `list.index` and `list.remove` time in time again to get the appropriate element in your names and data lists.